### ISSUE TITLE
libssh2 support as a runtime option instead of compile-time?

### ISSUE THREAD

**Issue by tkelman** [libgit2/libgit2#3762]
I have a use case where I'd like to make ssh support optional, where in a default configuration it is not available but if the libssh2 library becomes available at runtime I'd like libgit2 to be able to use it without having to recompile. Assuming I could ensure the versioning and build configuration of the runtime-available libssh2 matches what gets used at build time for libgit2, would this be at all technically feasible? Would it be possible to change some of the conditional compilation into some form of runtime check and associated API for initialization?

**Comment by ethomson** [comment]
Yes, this is definitely the avenue that we would like to pursue.  At the moment we're stuck in a terrible place where many consumers seem to want or expect SSH support, but building, releasing and maintaining libssh2 is _not_ something that we are prepared to do.  And asking the users of many of the bindings to recompile libgit2 is quite terrible.  (Most of them are either presented with a libgit2 binary, as with LibGit2Sharp, or their libgit2 is built for them, as with Rugged, and the end-user is not given any ability to tailor this behavior to even add the flags that indicate libssh2 support.)

My opinion is that using `dlopen` and `LoadLibraryEx` to resolve libssh2 is the superior solution here.  That way end-users of bindings, at least, can throw their own libssh2 down and we will load it, without having to recompile libgit2.

Now, this doesn't really fix _all_ the problems of distributing this to end-users, but it should (hopefully) help a bit.

**Comment by tiennou** [comment]
I'd be willing to work on that (the *nix part at least :trollface:). I'm wondering how to handle the "function call" part though... Can CMake do weak-linking ?

**Comment by tiennou** [comment]
Also, would it be acceptable to have a hard requirement for building with libssh2 or should the `GIT_SSH` tests be kept ?

**Comment by tkelman** [comment]
> would it be acceptable to have a hard requirement for building with libssh2

I think the way this would work is if libgit2 is present at build time, use it and link against it in the current way. If it isn't present at build time, then build some new runtime plugin stubs that aren't activated by default, or have to try `dlopen`ing a libssh at runtime and checking if it's compatible somehow. Maybe the build-time-only version could be made obsolete eventually, but since it's working now there's no reason to mess with it too much just yet.

**Comment by tiennou** [comment]
@tkelman Please have a look at the PR I opened ;-).
